x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java

```diff
@@ -771,14 +772,30 @@ static void writeDissect(PlanStreamOutput out, Dissect dissect) throws IOExcepti
     }
 
     static EsRelation readEsRelation(PlanStreamInput in) throws IOException {
-        return new EsRelation(in.readSource(), readEsIndex(in), readAttributes(in), in.readBoolean());
+        Source source = in.readSource();
+        EsIndex esIndex = readEsIndex(in);
+        List<Attribute> attributes = readAttributes(in);
+        EsSourceOptions esSourceOptions = null;
+        if (in.getTransportVersion().onOrAfter(TransportVersions.ESQL_ES_SOURCE_OPTIONS) && in.readBoolean()) {
+            esSourceOptions = new EsSourceOptions(in);
+        }
+        boolean frozen = in.readBoolean();
+        return new EsRelation(source, esIndex, attributes, esSourceOptions, frozen);
     }
 
     static void writeEsRelation(PlanStreamOutput out, EsRelation relation) throws IOException {
         assert relation.children().size() == 0;
         out.writeNoSource();
         writeEsIndex(out, relation.index());
         writeAttributes(out, relation.output());
+        if (out.getTransportVersion().onOrAfter(TransportVersions.ESQL_ES_SOURCE_OPTIONS)) {
+            if (relation.esSourceOptions() != null) {
+                out.writeBoolean(true);
```

Comment:
> Is marking the existence of the source option with a boolean really necessary? We encode the options as 3 optional strings

Without this boolean, how would the remote side know if it should attempt to read three (optional) strings or not? Reading an optional strings involves consuming at least a boolean.

Reply:
What I meant to say: `StreamOutput.writeOptionalString` already prepends a boolean to say "there's a string!" or "there's no string!". We're doubling this logic.

The serialization code in this PR, however, optimizes for the (common?) case where there's no options. My point is that this is unnecessary, because we'd only be sending 1 byte (`false`) instead of 3 bytes (`false, false, false`), at the cost of slightly more complex code (additional null-handling at the level of `EsSourceOptions`).

I don't feel strongly about this, but I'd prefer if
- the `EsSourceOptions` were not nullable (each of the options that are encoded in them is nullable already, so there's two ways to express that no options were set)
- and if the serialization always sent the `EsSourceOptions` as 3 optional strings, rather than specialcasing for when the `EsSourceOptions` is `null`.

Reply:
Got it, you were still addressing the nullability of the `EsSourceOptions` instance. Ok, took on your suggestion and applied it.